ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/PushRouting.java

```diff
@@ -238,13 +290,86 @@ public void close() {
+      final ConfiguredStatement<Query> statement,
+      final Set<KsqlNode> hosts,
+      final LogicalSchema outputSchema,
+      final TransientQueryQueue transientQueryQueue,
+      final PushConnectionsHandle pushConnectionsHandle
+  ) {
+    pushPhysicalPlan.getContext().runOnContext(v ->
+        checkForNewHosts(serviceContext, pushPhysicalPlan, statement, outputSchema,
+            transientQueryQueue, pushConnectionsHandle));
+  }
+
+  private void checkForNewHosts(
+      final ServiceContext serviceContext,
+      final PushPhysicalPlan pushPhysicalPlan,
+      final ConfiguredStatement<Query> statement,
+      final LogicalSchema outputSchema,
+      final TransientQueryQueue transientQueryQueue,
+      final PushConnectionsHandle pushConnectionsHandle
+  ) {
+    VertxUtils.checkContext(pushPhysicalPlan.getContext());
+    if (pushConnectionsHandle.isClosed()) {
+      return;
+    }
+    final Set<KsqlNode> updatedHosts = registryToNodes.apply(
+        pushPhysicalPlan.getScalablePushRegistry());
+    final Set<KsqlNode> hosts = pushConnectionsHandle.getAllHosts();
+    final Set<KsqlNode> newHosts = Sets.difference(updatedHosts, hosts).stream()
+        .filter(node -> !node.isLocal())
+        .collect(Collectors.toSet());
+    final Set<KsqlNode> removedHosts = Sets.difference(hosts, updatedHosts);
+    if (newHosts.size() > 0) {
+      LOG.info("Dynamically adding new hosts {}", newHosts);
+      connectToHosts(serviceContext, pushPhysicalPlan, statement, newHosts, outputSchema,
+          transientQueryQueue, pushConnectionsHandle, true);
+    }
+    if (removedHosts.size() > 0) {
+      LOG.info("Dynamically removing hosts {}", removedHosts);
+      for (final KsqlNode node : removedHosts) {
+        final RoutingResult result = pushConnectionsHandle.remove(node);
+        result.close();
+        result.updateStatus(RoutingResultStatus.REMOVED);
+      }
+    }
+    pushPhysicalPlan.getContext().owner().setTimer(clusterCheckInterval, timerId ->
+        checkForNewHosts(serviceContext, pushPhysicalPlan, statement, outputSchema,
+            transientQueryQueue, pushConnectionsHandle));
+  }
+
+  private static Set<KsqlNode> loadCurrentHosts(final ScalablePushRegistry scalablePushRegistry) {
+    return new HashSet<>(scalablePushRegistry
+        .getLocator()
+        .locate());
+  }
+
+  private static Function<ScalablePushRegistry, Set<KsqlNode>> createLoadingCache() {
+    final LoadingCache<ScalablePushRegistry, Set<KsqlNode>> cache = CacheBuilder.newBuilder()
+        .maximumSize(40)
+        .expireAfterWrite(2000, TimeUnit.MILLISECONDS)
+        .build(new CacheLoader<ScalablePushRegistry, Set<KsqlNode>>() {
+          @Override
+          public Set<KsqlNode> load(final ScalablePushRegistry scalablePushRegistry) {
+            return loadCurrentHosts(scalablePushRegistry);
+          }
+        });
+    return cache::getUnchecked;
+  }
+
   public enum RoutingResultStatus {
-    SUCCESS
+    IN_PROGRESS,
```

Comment:
nit: javadocs for each of these could help :)

Reply:
Done, added a bunch of javadoc.